### PR TITLE
Feature/atomikos

### PR DESCRIPTION
**Pull request by jimmarino**
As a follow-on to #541, this PR introduces support for transactions using the Atomikos transaction manager and JDBC data source connection pools. 

Atomikos supports both global (XA) and local transactions. Users can select which option is appropriate for their use case via data source driver configuration. 

Atomikos also provides robust connection pooling that is directly integrated with the transaction manager. Since Atomikos optimizes away the 2PC protocol when a single resource is enlisted, local transactions will not incur significant overhead with Atomikos enabled.

### CODE REVIEW

**Review by denisneuling — CHANGES_REQUESTED** [review]
Hi Jim,

many thanks for supplying the Atomikos TX manager integration. 

Do you think we can make it capable of wrapping registered datasources instead of leaving datasource creation to it so we can take advantage of our own `org.eclipse.dataspaceconnector.spi.transaction.datasource.DataSourceRegistry#register(String name, javax.sql.DataSource ds)` spi?

Afaik one is able to wrap `javax.sql.XADataSource`s with Atomikos in the `org.eclipse.dataspaceconnector.transaction.atomikos.AtomikosDataSourceRegistry` like so:

```java
public void register(String name, DataSource dataSource) {
    if(dataSource instanceof XADataSource) {
        AtomikosDataSourceBean bean = new AtomikosDataSourceBean();
        bean.setXaDataSource((XADataSource) dataSource);
        dataSource = bean;
    }
    dataSources.put(name, dataSource);
}
```

Unfortunately I couldn't find out how to do so with the `com.atomikos.jdbc.AtomikosNonXADataSourceBean`. Obviously a non-XA datasource needs to be made aware of the tx manager to become scoped by it.
Dou you have an idea how to accomplish that without introducing an own initialization functionatliy to the AtomikosDataSourceRegistry?

All the rest looks very promising :+1: 

<sub> Denis Neuling <denis.neuling@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)

**Review by paullatzelsperger — APPROVED** [review]
I just saw that Atomikos seems to either want to have SLF4J's  `StaticLoggerBinder`, or a Log4J `Logger` on the classpath. If it finds none, it's just gonna vomit logs all over the Console using their fallback `JULLogger`.
Another way of doing it would be to register a `Logger` with the `"com.atomikos.logging.LoggerFactory" name, effectively overwriting the default one provided by Atomikos. Could quite figure out how to do that properly though. 

TL;DR, we should do this in another PR IMO.

**Comment by jimmarino** [comment]
> Hi Jim,
> 
> many thanks for supplying the Atomikos TX manager integration.
> 
> Do you think we can make it capable of wrapping registered datasources instead of leaving datasource creation to it so we can take advantage of our own `org.eclipse.dataspaceconnector.spi.transaction.datasource.DataSourceRegistry#register(String name, javax.sql.DataSource ds)` spi?
> 
> Afaik one is able to wrap `javax.sql.XADataSource`s with Atomikos in the `org.eclipse.dataspaceconnector.transaction.atomikos.AtomikosDataSourceRegistry` like so:
> 
> ```java
> public void register(String name, DataSource dataSource) {
>     if(dataSource instanceof XADataSource) {
>         AtomikosDataSourceBean bean = new AtomikosDataSourceBean();
>         bean.setXaDataSource((XADataSource) dataSource);
>         dataSource = bean;
>     }
>     dataSources.put(name, dataSource);
> }
> ```
> 
> Unfortunately I couldn't find out how to do so with the `com.atomikos.jdbc.AtomikosNonXADataSourceBean`. Obviously a non-XA datasource needs also to be made aware of the tx manager to become scoped by it. Dou you have an idea how to accomplish that without introducing an own initialization functionatliy to the `AtomikosDataSourceRegistry`?
> 
> All the rest looks very promising 👍
> 
> Denis Neuling [denis.neuling@daimler.com](mailto:denis.neuling@daimler.com), Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)

Good question. I'm not sure that is possible. But we can support the XA case. Should we do that in a subsequent PR?

**Comment by denisneuling** [comment]
Alright, let's try to find this out in another PR. Good to merge from my side :+1:

**Review by denisneuling — APPROVED** [review]

File: extensions/transaction/transaction-atomikos/src/main/java/org/eclipse/dataspaceconnector/transaction/atomikos/AtomikosTransactionExtension.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import static org.eclipse.dataspaceconnector.transaction.atomikos.TransactionManagerConfigurationKeys.DATA_DIR;
+import static org.eclipse.dataspaceconnector.transaction.atomikos.TransactionManagerConfigurationKeys.LOGGING;
+import static org.eclipse.dataspaceconnector.transaction.atomikos.TransactionManagerConfigurationKeys.THREADED2PC;
+import static org.eclipse.dataspaceconnector.transaction.atomikos.TransactionManagerConfigurationKeys.TIMEOUT;
+
+
+/**
+ * Provides an implementation of a {@link DataSourceRegistry} and a {@link TransactionContext} backed by Atomikos.
+ */
+public class AtomikosTransactionExtension implements ServiceExtension {
```
[the remark this answers is not in the document]
Comment (jimmarino) [reply extensions/transaction/transaction-atomikos/src/main/java/org/eclipse/dataspaceconnector/transaction/atomikos/AtomikosTransactionExtension.java:38]: Ah, good catch, I forgot that

File: extensions/transaction/transaction-atomikos/src/main/java/org/eclipse/dataspaceconnector/transaction/atomikos/AtomikosTransactionContext.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+
+import static javax.transaction.Status.STATUS_ACTIVE;
+import static javax.transaction.Status.STATUS_MARKED_ROLLBACK;
+
+/**
+ * An implementation backed by the Atomikos transaction manager.
+ */
+public class AtomikosTransactionContext implements TransactionContext {
```
[the remark this answers is not in the document]
Comment (jimmarino) [reply extensions/transaction/transaction-atomikos/src/main/java/org/eclipse/dataspaceconnector/transaction/atomikos/AtomikosTransactionContext.java:32]: Yes, I decided not to do that for the time being since there is not another impl, but we can always pull that out.

**Comment by jimmarino** [comment]
> Alright, let's try to find this out in another PR. Good to merge from my side 👍

Thx!

**Review by ndr-brt — APPROVED** [review]
